libr/io/io_map.c

```diff
@@ -493,3 +450,187 @@ R_API RIOMap *r_io_map_get_by_ref(RIO *io, RIOMapRef *ref) {
+	if (!node) {
+		return;
+	}
+	MapOverlayChunk *chunk = NULL;
+	RRBNode *prev = r_rbnode_prev (node);
+	while (prev) {
+		chunk = (MapOverlayChunk *)prev->data;
+		if (!r_itv_overlap (chunk->itv, search_itv)) {
+			break;
+		}
+		node = prev;
+		prev = r_rbnode_prev (prev);
+	}
+	chunk = (MapOverlayChunk *)node->data;
+	do {
+		addr = R_MAX (r_itv_begin (search_itv), r_itv_begin (chunk->itv));
+		ut8 *dst = &buf[addr - r_itv_begin (search_itv)];
+		ut8 *src = &chunk->buf[addr - r_itv_begin (chunk->itv)];
+		memcpy (dst, src, (size_t)(R_MIN (r_itv_end (search_itv), r_itv_end (chunk->itv)) - addr));
+		node = r_rbnode_next (node);
+		chunk = node? (MapOverlayChunk *)node->data: NULL;
+	} while (chunk && r_itv_overlap (chunk->itv, search_itv));
+}
+
+static void _overlay_chunk_free(void *data) {
+	if (!data) {
+		return;
+	}
+	MapOverlayChunk *chunk = (MapOverlayChunk *)data;
+	free (chunk->buf);
+	free (chunk);
+}
+
+static int _overlay_chunk_insert (void *incoming, void *in, void *user) {
+	MapOverlayChunk *incoming_chunk = (MapOverlayChunk *)incoming;
+	MapOverlayChunk *in_chunk = (MapOverlayChunk *)in;
+	if (r_itv_begin (incoming_chunk->itv) < r_itv_begin (in_chunk->itv)) {
+		return -1;
+	}
+	if (r_itv_begin (incoming_chunk->itv) > r_itv_begin (in_chunk->itv)) {
+		return 1;
+	}
+	return 0;
+}
+
+R_API bool r_io_map_write_to_overlay(RIOMap *map, ut64 addr, const ut8 *buf, int len) {
+	r_return_val_if_fail (map && buf, false);
+	RInterval x = {addr, len};
+	RInterval search_itv = r_itv_intersect (map->itv, x);
+	if (!r_itv_size (search_itv)) {
+		return true;	// is this correct?
+	}
+	if (!map->overlay) {
+		if (!(map->overlay = r_crbtree_new (_overlay_chunk_free))) {
+			return false;
+		}
+	}
+	search_itv.addr -= map->itv.addr;
+	RRBNode *node = r_crbtree_find_node (map->overlay, &search_itv, _overlay_chunk_find, NULL);
+	if (!node) {
+		MapOverlayChunk *chunk = R_NEW0 (MapOverlayChunk);
+		if (!chunk) {
+			return false;
+		}
+		chunk->buf = R_NEWS (ut8, r_itv_size (search_itv));
+		chunk->itv = search_itv;
+		if (!chunk->buf || !r_crbtree_insert (map->overlay, chunk, _overlay_chunk_insert, NULL)) {
+			free (chunk->buf);
+			free (chunk);
+			return false;
+		}
+		memcpy (chunk->buf, buf, r_itv_size (search_itv));
+		return true;
+	}
+	MapOverlayChunk *chunk = NULL;
+	RRBNode *prev = r_rbnode_prev (node);
+	while (prev) {
+		chunk = (MapOverlayChunk *)prev->data;
+		if (!r_itv_overlap (chunk->itv, search_itv)) {
+			break;
+		}
+		node = prev;
+		prev = r_rbnode_prev (prev);
+	}
+	chunk = (MapOverlayChunk *)node->data;
+	if (r_itv_include (chunk->itv, search_itv)) {
+		ut8 *dst = &chunk->buf[r_itv_begin (search_itv) - r_itv_begin (chunk->itv)];
+		memcpy (dst, buf, r_itv_size (search_itv));
+		return true;
+	}
+	if (r_itv_begin (chunk->itv) < r_itv_begin (search_itv)) {
+		chunk->itv.size = r_itv_begin (search_itv) - r_itv_begin (chunk->itv);
+		chunk->buf = realloc (chunk->buf, r_itv_size (chunk->itv));
```

Comment:
check if realloc returns valid pointer

Reply:
reloc always returns a valid pointer here, because the new size is guaranteed to be smaller than the old size